File: fnnls/fnnls.py
# FNNLSa implementation from Rasmus Bro (1997) fnnls.m (MATLAB file exchange)
# translated to Python
# Daniel Elnatan
# Jan 22, 2018
# I believe I fixed the following fnnls algorithm by solving an indexing error.
# Ian Anthony
# August 8, 2019

from numpy import (
    zeros,
    arange,
    int64,
    float64,
    sum,
    argmax,
    nonzero,
    diag,
    min,
    abs,
    newaxis,
    finfo,
)
from scipy.linalg import solve

nu = newaxis
import numpy as np
import sys
from scipy.optimize import nnls

# machine epsilon
eps = finfo(float64).eps


def any(a):
    # assuming a vector, a
    larger_than_zero = sum(a > 0)
    if larger_than_zero:
        return True
    else:
        return False


def find_nonzero(a):
    # returns indices of nonzero elements in a
    return nonzero(a)[0]


def FNNLSa(XtX, Xty):
    tol = eps

    M, N = XtX.shape
    # initialize passive set, P. Indices where coefficient is >0
    P = zeros(N, dtype=int64)
    # and active set. Indices where coefficient is <=0
    Z = arange(N) + 1
    # working active set
    ZZ = arange(N) + 1
    # initial solution vector, x
    x = zeros(N, dtype=float64)
    # weight vector
    w = Xty - XtX @ x
    # iteration counts and parameter
    it = 0
    itmax = 30 * N

    # MAIN LOOP
    # continue as long as there are indices within the active set Z
    # or elements in inner loop active set is larger than 'tolerance'
    piter = 0
    while any(Z) and any(w[ZZ - 1] > tol):
        piter += 1
        t = argmax(w[ZZ - 1]) + 1  # find largest weight
        t = ZZ[t - 1]
        P[t - 1] = t  # move to passive set
        Z[t - 1] = 0  # remove from active set
        PP = find_nonzero(P) + 1
        ZZ = find_nonzero(Z) + 1
        NZZ = ZZ.shape

        # compute trial solution, s
        s = zeros(N, dtype=float64)

        if len(PP) == 1:
            s[PP - 1] = Xty[PP - 1] / XtX[PP - 1, PP - 1]
        else:
            s[PP - 1] = solve(XtX[PP - 1, PP[:, nu] - 1], Xty[PP - 1])
        s[ZZ - 1] = 0.0  # set active coefficients to 0

        while any(s[PP - 1] <= tol) and it < itmax:
            it = it + 1
            QQ = find_nonzero((s <= tol) * P) + 1
            alpha = min(x[QQ - 1] / (x[QQ - 1] - s[QQ - 1]))
            x = x + alpha * (s - x)
            ij = find_nonzero((abs(x) < tol) * (P != 0)) + 1
            Z[ij - 1] = ij
            P[ij - 1] = 0
            PP = find_nonzero(P) + 1
            ZZ = find_nonzero(Z) + 1
            if len(PP) == 1:
                s[PP - 1] = Xty[PP - 1] / XtX[PP - 1, PP - 1]
            else:
                s[PP - 1] = solve(XtX[PP - 1, PP[:, nu] - 1], Xty[PP - 1])
            s[ZZ - 1] = 0.0
        # assign current solution, s, to x
        x = s
        # recompute weights
        w = Xty - XtX @ x

    return x, w


XtX = np.array(
    [
        [
            9.611026155476322,
            8.87111685589437,
            8.211011366377516,
            7.621082704720055,
            7.092939797905212,
            6.619262476847601,
            6.193659100250106,
            5.810543643147376,
        ],
        [
            8.87111685589437,
            8.211011366377514,
            7.621082704720055,
            7.092939797905212,
            6.619262476847601,
            6.193659100250106,
            5.810543643147375,
            5.465029534777697,
        ],
        [
            8.211011366377516,
            7.621082704720055,
            7.092939797905212,
            6.619262476847601,
            6.193659100250106,
            5.810543643147375,
            5.465029534777699,
            5.152837915031511,
        ],
        [
            7.621082704720055,
            7.092939797905212,
            6.619262476847601,
            6.193659100250106,
            5.810543643147375,
            5.465029534777699,
            5.152837915031511,
            4.8702183082070425,
        ],
        [
            7.092939797905212,
            6.619262476847601,
            6.193659100250106,
            5.810543643147375,
            5.465029534777698,
            5.152837915031511,
            4.8702183082070425,
            4.613879995136789,
        ],
        [
            6.619262476847601,
            6.193659100250106,
            5.810543643147375,
            5.465029534777699,
            5.152837915031511,
            4.8702183082070425,
            4.613879995136789,
            4.380932606745861,
        ],
        [
            6.193659100250106,
            5.810543643147375,
            5.465029534777699,
            5.152837915031511,
            4.8702183082070425,
            4.613879995136789,
            4.380932606745861,
            4.1688346695848635,
        ],
        [
            5.810543643147376,
            5.465029534777697,
            5.152837915031511,
            4.8702183082070425,
            4.613879995136789,
            4.380932606745861,
            4.1688346695848635,
            3.975349011820167,
        ],
    ]
)
Xty = np.array(
    [
        6.292528486334608,
        5.899123987164679,
        5.54451856447887,
        5.224283191432115,
        4.934536993948907,
        4.671875931227347,
        4.433311073124163,
        4.2162151516528645,
    ]
)

print(FNNLSa(XtX, Xty), "\n")

print(nnls(XtX, Xty))

File: fnnls/src/fnnls.rs
#![allow(non_snake_case)]

use ndarray::{prelude::*, Dimension};
use ndarray_linalg::Solve;
use std::f64::EPSILON;

pub fn fnnls(xtx: &Array2<f64>, xty: &Array1<f64>) -> (Array1<f64>, Array1<f64>) {
    let (M, N) = (xtx.nrows(), xtx.ncols());
    let mut P = Array::zeros(M); // passive; indices with vals > 0
    let mut Z = Array::from_iter(0..N) + 1; // active; i w/ vals <= 0
    let mut ZZ = Array::from_iter(0..N) + 1; // working active set
    let mut x: Array1<f64> = Array::zeros(N); // initial solution vector
    let mut w = xty - &(xtx.dot(&x)); // weight vector
    let mut it = 0; // iterator for the while loop
    let itmax = 30 * N; // maximum iterations

    // Continue if indices in the active set or values > than machine epsilon
    while Z.iter().any(|&i| i > 0) && ZZ.iter().any(|&i| w[i - 1] > EPSILON) {
        let t = max_index(&(ZZ.mapv(|i| w[i - 1]))) + 1;
        P[ZZ[t - 1] - 1] = ZZ[t - 1]; // move to the passive set
        Z[ZZ[t - 1] - 1] = 0; // remove from the active set
        ZZ = Array::from(find_nonzero(&Z)) + 1;
        let mut PP = Array::from(find_nonzero(&P)) + 1;
        let mut PPcopy = Array::from(find_nonzero(&P)) + 1;
        let mut s: Array1<f64> = Array::zeros(N); // trial solution
        match PP.len() {
            0 => s[0] = 0.0,
            1 => s[PP[0] - 1] = xty[PP[0] - 1] / xtx[[PP[0] - 1, PP[0] - 1]],
            _ => {
                let xtx_pp_solution = slice_with_array(xtx, &(PPcopy - 1))
                    .solve_into(PP.mapv(|i| xty[i - 1]))
                    .unwrap(); // solve PP-reduced set of xtx @ xty
                for (i, &value) in PP.indexed_iter() {
                    s[value - 1] = xtx_pp_solution[i];
                }
            }
        }
        for &i in ZZ.iter() {
            s[i - 1] = 0.0; // set active coefficients to 0
        }

        while PP.iter().any(|&i| s[i - 1] <= EPSILON) && it < itmax {
            it += 1;
            let s_mask = s.mapv(|e| e <= EPSILON);
            let tmp = P
                .indexed_iter()
                .map(|(i, &v)| if s_mask[[i]] { v } else { 0 })
                .collect::<Vec<usize>>();
            let QQ = Array::from(find_nonzero_vec(&tmp)) + 1;
            let xQQ = QQ.mapv(|i| x[i - 1]);
            let alpha = min(&(&xQQ / &(&xQQ - &QQ.mapv(|i| s[i - 1]))));
            x = &x + &(alpha * (&s - &x));
            let mask = P.mapv(|i| i != 0) & x.mapv(|i| i.abs() < EPSILON);
            for (i, &v) in mask.indexed_iter() {
                if v {
                    Z[i] = i + 1;
                    P[i] = 0;
                }
            }
            PP = Array::from(find_nonzero(&P)) + 1;
            PPcopy = Array::from(find_nonzero(&P)) + 1;
            ZZ = Array::from(find_nonzero(&Z)) + 1;
            match PP.len() {
                // verbatim repeat of the previous match statement
                0 => s[0] = 0.0,
                1 => s[PP[0] - 1] = xty[PP[0] - 1] / xtx[[PP[0] - 1, PP[0] - 1]],
                _ => {
                    let xtx_pp_solution = slice_with_array(xtx, &(PPcopy - 1))
                        .solve_into(PP.mapv(|i| xty[i - 1]))
                        .unwrap();
                    for (i, &value) in PP.indexed_iter() {
                        s[value - 1] = xtx_pp_solution[i];
                    }
                }
            }
            for &i in ZZ.iter() {
                s[i - 1] = 0.0;
            }
        }
        x = s; // assign current solution (s) to x
        w = xty - &(xtx.dot(&x)); // recompute weights
    }
    (x, w)
}

pub fn find_nonzero_vec(vec: &[usize]) -> Vec<usize> {
    vec.iter()
        .enumerate()
        .filter(|(_, &value)| value != 0)
        .map(|(i, _)| i)
        .collect::<Vec<usize>>()
}

pub fn find_nonzero(array: &Array1<usize>) -> Vec<usize> {
    array
        .indexed_iter()
        .filter(|(_, &value)| value != 0)
        .map(|(i, _)| i)
        .collect::<Vec<usize>>()
}

pub fn slice_with_array(a: &Array2<f64>, b: &Array1<usize>) -> Array2<f64> {
    let mut newArray = Array::zeros((b.len(), b.len()));
    for i in 0..b.len() {
        for j in 0..b.len() {
            newArray[[i, j]] = a[[b[j], b[i]]]
        }
    }
    newArray
}

pub fn max_index(array: &Array1<f64>) -> usize {
    let mut index = 0;
    for (i, &value) in array.iter().enumerate() {
        if value > array[index] {
            index = i;
        }
    }
    index
}

pub fn min<D: Dimension>(array: &Array<f64, D>) -> f64 {
    array.iter().cloned().fold(1.0 / 0.0, f64::min)
}

pub fn test_setup() -> (Array2<f64>, Array1<f64>) {
    let random_array_a = array![
        [
            9.611_026_155_476_322,
            8.871_116_855_894_37,
            8.211_011_366_377_516,
            7.621_082_704_720_055,
            7.092_939_797_905_212,
            6.619_262_476_847_601,
            6.193_659_100_250_106,
            5.810_543_643_147_376
        ],
        [
            8.871_116_855_894_37,
            8.211_011_366_377_514,
            7.621_082_704_720_055,
            7.092_939_797_905_212,
            6.619_262_476_847_601,
            6.193_659_100_250_106,
            5.810_543_643_147_375,
            5.465_029_534_777_697
        ],
        [
            8.211_011_366_377_516,
            7.621_082_704_720_055,
            7.092_939_797_905_212,
            6.619_262_476_847_601,
            6.193_659_100_250_106,
            5.810_543_643_147_375,
            5.465_029_534_777_699,
            5.152_837_915_031_511
        ],
        [
            7.621_082_704_720_055,
            7.092_939_797_905_212,
            6.619_262_476_847_601,
            6.193_659_100_250_106,
            5.810_543_643_147_375,
            5.465_029_534_777_699,
            5.152_837_915_031_511,
            4.870_218_308_207_042_5
        ],
        [
            7.092_939_797_905_212,
            6.619_262_476_847_601,
            6.193_659_100_250_106,
            5.810_543_643_147_375,
            5.465_029_534_777_698,
            5.152_837_915_031_511,
            4.870_218_308_207_042_5,
            4.613_879_995_136_789
        ],
        [
            6.619_262_476_847_601,
            6.193_659_100_250_106,
            5.810_543_643_147_375,
            5.465_029_534_777_699,
            5.152_837_915_031_511,
            4.870_218_308_207_042_5,
            4.613_879_995_136_789,
            4.380_932_606_745_861
        ],
        [
            6.193_659_100_250_106,
            5.810_543_643_147_375,
            5.465_029_534_777_699,
            5.152_837_915_031_511,
            4.870_218_308_207_042_5,
            4.613_879_995_136_789,
            4.380_932_606_745_861,
            4.168_834_669_584_863_5
        ],
        [
            5.810_543_643_147_376,
            5.465_029_534_777_697,
            5.152_837_915_031_511,
            4.870_218_308_207_042_5,
            4.613_879_995_136_789,
            4.380_932_606_745_861,
            4.168_834_669_584_863_5,
            3.975_349_011_820_167
        ]
    ];

    let random_array_b = array![
        6.292_528_486_334_608,
        5.899_123_987_164_679,
        5.544_518_564_478_87,
        5.224_283_191_432_115,
        4.934_536_993_948_907,
        4.671_875_931_227_347,
        4.433_311_073_124_163,
        4.216_215_151_652_864_5
    ];

    (random_array_a, random_array_b)
}
